reference.md

```diff
@@ -244,6 +244,39 @@ Example:
   (table.concat c ",")) ; => "3,4,5,6"
 ```
 
+### `let-open` bind and auto-close file handles
+
+*(Since ??? TODO: add version before release)*
+
+While Lua will automatically close an open file handle when it's garbage collected,
+GC may not run right away; `let-open` ensures handles are closed immediately, error
+or no, without boilerplate.
+
+The usage is the same as `let`, only every binding should be a file handle or other value
+with a `:close` method. After executing the body, or upon encountering an error, `let-open`
+will invoke `(value:close)` on every bound variable before returning the results.
+
+The body is implicitly wrapped in a function and run with `xpcall` so that all bound
+handles are closed before it re-raises the error.
+
+Example:
+
+```fennel
+; Basic usage
+(let-open [fout (io.open :output.txt :w) fin (io.open :input.txt)]
+  (fout:write "Here is some text!\n")
+  (fin:line)) ; => first line of input.txt
+
+; This demonstrates that the file will also be closed upon error.
+(var fh nil)
+(local (ok err)
+  (pcall #(let-open [file (io.open :test.txt :w)]
+            (set fh file) ; you would normally never do this
+            (error :whoops!))))
+(io.type fh) ; => "closed file"
```

Comment:
Do you think maybe we could show trying to write to the file here and show that it has an error instead?

Reply:
You mean assume the write errored instead of causing an error in the code?

Reply:
Yeah, show that attempting to write to `file` results in an error outside the `let-open` form maybe? Just an idea since it might be clearer than the `io.type` function; that might be less familiar to folks, but it's fine as-is too.

Reply:
Hrm, well the `io.type` is there mainly to demonstrate that even when there *are* errors, the file-handlers are still closed; demonstrating that the errors still propagate was more of a bonus.

I'm happy to change it that makes these points clearer, but my gut says we should still demonstrate both behaviors. Any ideas that would demonstrate both with more clarity? Off the top of my head, instead of explicitly raising an error I could have something like `(file:nonexisted-method "this will raise an error")`.

We could always merge it as-is and clarify the reference later, too.

Reply:
Yeah, we can change this after merging. Everything else looks great. Nice work.